packages/core/src/core/deploy_summary.ts

```
@@ -0,0 +1,175 @@
+* limitations under the License.
+*/
+
+/* eslint-disable default-case */
+/* eslint-disable no-fallthrough */
+/* eslint-disable consistent-return */
+
+import { Change,
+  isAdditionOrRemovalChange,
+  getChangeElement,
+  isAdditionChange,
+  isRemovalOrModificationChange,
+  DetailedChange,
+  ChangeDataType } from '@salto-io/adapter-api'
+import { detailedCompare } from '@salto-io/adapter-utils'
+import { AdditionDiff, ModificationDiff, RemovalDiff } from '@salto-io/dag'
+import _ from 'lodash'
+
+export type DeployResult = 'success'|'failure'|'partial-success'
+export type DetailedChangeId = string
+export type DetailedChangeDeployResult = [DetailedChangeId, DeployResult]
+
+const handleEmptyAppliedChange = (
+  requestedChange: Change,
+): DetailedChangeDeployResult[] => {
+  const requestedChangeName = getChangeElement(requestedChange).elemID.getFullName()
+  if (isAdditionOrRemovalChange(requestedChange)) {
+    return [[requestedChangeName, 'failure']]
+  }
+  const requestedDetailedChanges = detailedCompare(
+    requestedChange.data.before,
+    requestedChange.data.after
+  )
+  return requestedDetailedChanges
+    .map(detailedChange => [detailedChange.id.getFullName(), 'failure'])
+}
+
+const summarizeAdditionChange = (
+  requestedChange: AdditionDiff<ChangeDataType>,
+  appliedChange: Change
+): DetailedChangeDeployResult[] => {
+  const requestedChangeName = getChangeElement(requestedChange).elemID.getFullName()
+
+  switch (appliedChange.action) {
+    case 'add': {
+      const diffBetweenRequestedAndApplied = detailedCompare(
+        requestedChange.data.after,
+        appliedChange.data.after
+      )
+      const deployResult: DeployResult = diffBetweenRequestedAndApplied.every(isAdditionChange)
+        ? 'success'
+        : 'partial-success'
+      return [[requestedChangeName, deployResult]]
+    }
+    case 'remove':
+    case 'modify':
+    // Impossible cases, fail them
```

I wondered how to address these cases, I debated between 3 solutions here:

* Return an Empty result `[[]]`  \ Error Result ( `id: failure` )  \ Throw an error (`InvalidStateReached`)

I think the "Empty Result" solution is invalid here - since this function should give as much information as possible, and no reason not to say anything, when we have something to say ( aka, something "Bad" happened)

I think the Throw an error solution is also "weird" since I prefer to write "pure" functions ( which are easier to reason about their behavior and output) to outside callers ( I wouldn't expect this function to throw an error, and even further, I don't really care when this logical error happens as a consumer, since I only want to understand what went wrong... )

The last solution, is just return a "failure" on the request change ( which Is what I did )

WDYT?

---

returning a failure seems like the correct logic - we don't expect this to happen, but if it does, saying that something failed seems like a fair representation of the situation.
only thing I would add here is an error log, I think it is a good habit to have these anywhere there is a "should never happen" comment that doesn't crash because it makes it a bit more visible when those scenarios end up happening anyway

---

👍 added a log.
The severity I added is "Warn".
LMK if you disagree ( and this should be "Error" in your opinion)